### PR TITLE
Add setting to read all packets until timeout expires.

### PR DESCRIPTION
**Pull request by maffoo**
Discussed in #276 that this could be used to simplify the board autodetection logic.

### CODE REVIEW

**Comment by maffoo** [comment]
Note that this would be the first API incompatibility between the old delphi and new scala direct ethernet servers. So if we change the ghz_fpga_server to use this, there'd be no going back.

**Comment by jwenner** [comment]
With the API incompatibility, there should be either be a note in the docstring (both internal and external, if different) or a scala warning (if possible) warning users about the incompatibility. This is to ensure that someone in the future doesn't use this function and accidentally break Delphi compatibility.

Further, at such a time, it might be appropriate to replace the Delphi Direct Ethernet server with something that just raises an Exception saying that the server is deprecated and to use the scala direct ethernet server instead.

**Comment by maffoo** [comment]
@jwenner, I added a note in the docstring (which will be visible to labrad clients) about when the new setting was added.

There's not really a way to deprecate a server; I think our best bet will be to just remove the Delphi binary from the repo. To my knowledge all of our setups have moved to the scala server and no one has used the old delphi version in quite some time.

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala [hunk 1]
```diff
@@ -180,6 +180,17 @@ extends ServerContext with Logging {
     }
   }
 
+  private def processPackets(packets: Seq[EthernetPacket]): Seq[(String, String, Int, Array[Byte])] = {
```
Comment (DanielSank) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: What's our stand on docs?

Comment (maffoo) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: docs are good, but in my mind the name and type signature here is really all you need to know. More text here would seem to me to just repeat what's already pretty evident in the code.

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala [hunk 2]
```diff
@@ -190,15 +201,7 @@ extends ServerContext with Logging {
...
-    val results = queue.take(numPackets.toInt)
-    results.map { pkt =>
-      val src = pkt.getHeader.getSrcAddr
-      val dst = pkt.getHeader.getDstAddr
-      val typ = pkt.getHeader.getType.value.toInt
-      val data = pkt.getPayload.getRawData
-
-      (src.toString.toUpperCase, dst.toString.toUpperCase, typ, data)
-    }
+    processPackets(queue.take(numPackets.toInt))
```
Comment (DanielSank) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: Isn't converting Long to Int kinda janky?

**Comment by jwenner** [comment]
For deprecation, shouldn't we be able to replace the Delphi direct ethernet exe with a python server which identifies itself as "Direct Ethernet" (just like the Delphi version)? Then have only initServer, where we would raise an Exception with a message about the server being deprecated and to use the scala server instead.

**Comment by maffoo** [comment]
@jwenner, yes, we could do something like that. I just don't think it's worth it. Anyone who needs the old delphi version can get it from the repo history. The rest of us can move on.

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala [hunk 3]
```diff
@@ -228,6 +231,24 @@ extends ServerContext with Logging {
...
+            |
+            |Unlike Collect, Read and Discard, this is useful in cases where we
+            |do not know how many packets are expected, but just wish to capture
+            |all traffic in a given interval.
+            |
+            |New in version 1.3.0; not present in Delphi direct ethernet server.""")
+  def readUntilTimeout(): Seq[(String, String, Int, Array[Byte])] = {
+    val timeout = readTimeout.getOrElse { fail(11, "Read Timeout not set") }
+    val millis = timeout.toMillis
+    val nanos = (timeout - millis.milliseconds).toNanos.toInt
```
Comment (DanielSank) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: Wat?

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala (hunk 2 above)
Comment (maffoo) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: If you are asking for more that 2 billion packets, you have bigger problems. We could add a check, but I'm not too worried about it. Using `Long` as a proxy for `unsigned 32-bit int` is dumb, but it's going to take a lot of work to fix that.

**Comment by jwenner** [comment]
@maffoo, the reason would be because other groups seem to be using the FPGA server (and thus I would assume the Direct Ethernet server), so such a deprecation would be a notice to such external users.

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala [hunk 4]
```diff
@@ -228,6 +231,24 @@ extends ServerContext with Logging {
   }
 
   @Setting(
+    id = 53,
+    name = "Read until timeout",
+    doc = """Collect and return all packets that arrive until the timeout expires
+            |
+            |Unlike Collect, Read and Discard, this is useful in cases where we
```
Comment (DanielSank) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: I think we need a good ol' Oxford comma after `Read`.

File: DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala (hunk 2 above)
Comment (DanielSank) [inline DirectEthernet/src/main/scala/org/labrad/ethernet/server/DirectEthernet.scala]: Give the possibility for weird behavior, I'd add the check.

**Comment by DanielSank** [comment]
> @jwenner, yes, we could do something like that. I just don't think it's worth it. Anyone who needs the old delphi version can get it from the repo history. The rest of us can move on.

Agreed. If people need the old direct ethernet server it's because they're using the old fpga server. If people want to stick with and/or fork from old code, that's entirely their decision, and not one we have any reason to support.

@maffoo LGTM but it would be a good idea to check that this works on hardware. Is that what you're doing on redheatdos?

**Comment by maffoo** [comment]
@DanielSank; I was checking the new pcap4j library on redheatdos. We're not yet using the new setting added here, but I could modify the fpga server to use it and test that if you want to see it working.

**Comment by DanielSank** [comment]
Unfortunately I'm away and will be travelling a lot so it's a bit hard for me to test stuff (although I guess you're used to this workflow so maybe you have wisdom). In any case, what would you say to doing the FPGA work that uses this in the same branch and PRing that stuff all-together? At least that way we get to see that it works right.

Alternatively we could set up a real unit test.

**Pull request closed without merging**